deltacat/compute/compactor_v2/steps/merge.py

```diff
@@ -345,23 +483,44 @@ def _timed_merge(input: MergeInput) -> MergeResult:
         f"merge_{worker_id}_{task_id}.bin"
     ) if input.enable_profiler else nullcontext():
         total_input_records, total_deduped_records = 0, 0
+        total_dropped_records = 0
         materialized_results: List[MaterializeResult] = []
         merge_file_groups = input.merge_file_groups_provider.create()
         hb_index_copy_by_ref_ids = []
 
         for merge_file_group in merge_file_groups:
-            if not merge_file_group.dfe_groups:
+            compacted_table = None
+            has_delete = input.delete_file_envelopes is not None
+            if has_delete:
+                assert (
+                    input.delete_strategy is not None
+                ), "Merge input missing delete_strategy"
+            if not has_delete and not merge_file_group.dfe_groups:
+                # Can copy by reference only if there are no deletes to merge in
                 hb_index_copy_by_ref_ids.append(merge_file_group.hb_index)
                 continue
-
-            table, input_records, deduped_records = _compact_tables(
-                input, merge_file_group.dfe_groups, merge_file_group.hb_index
+            if _has_previous_compacted_table(input, merge_file_group.hb_index):
+                compacted_table = _download_compacted_table(
+                    hb_index=merge_file_group.hb_index,
+                    rcf=input.round_completion_info,
+                    read_kwargs_provider=input.read_kwargs_provider,
+                    deltacat_storage=input.deltacat_storage,
+                    deltacat_storage_kwargs=input.deltacat_storage_kwargs,
+                )
+            table, input_records, deduped_records, dropped_records = _compact_table_v2(
+                input,
+                merge_file_group.dfe_groups,
+                merge_file_group.hb_index,
+                has_delete,
+                compacted_table,
             )
             total_input_records += input_records
             total_deduped_records += deduped_records
-            materialized_results.append(
-                merge_utils.materialize(input, merge_file_group.hb_index, [table])
-            )
+            total_dropped_records += dropped_records
+            if table is not None:
```

Comment:
will table ever be None at this point?

Reply:
Fixed now - the resulting table would be None if `merge_file_group.dfe_groups` is `None` and `compacted_table` is `None`